File: scripts/probe_sources.py
"""Inspect an imported collection and sample live sources with the production Lua parser.

Transport/encoding use Python on the development PC, not Kindle LuaSocket/iconv.
No source JavaScript is executed. Reports omit response text and credentials.
"""

from __future__ import annotations

import argparse
from collections import Counter
from datetime import datetime, timezone
import hashlib
import ipaddress
import json
import os
from pathlib import Path
import socket
import time
import urllib.error
import urllib.parse
import urllib.request

from lupa.luajit21 import LuaRuntime

ROOT = Path(__file__).resolve().parent.parent


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, *args):
        return None


def safe_site(url: str) -> str:
    try:
        parsed = urllib.parse.urlsplit(url)
        return f"{parsed.scheme}://{parsed.hostname or ''}"
    except ValueError:
        return "invalid"


def public_url(url: str) -> str:
    parsed = urllib.parse.urlsplit(url)
    if parsed.scheme not in {"http", "https"} or not parsed.hostname or parsed.username or parsed.password:
        raise ValueError("invalid public HTTP URL")
    addresses = socket.getaddrinfo(parsed.hostname, parsed.port or (443 if parsed.scheme == "https" else 80))
    if not addresses or any(not ipaddress.ip_address(item[4][0]).is_global for item in addresses):
        raise ValueError("non-public address")
    return urllib.parse.quote(urllib.parse.urldefrag(url)[0], safe=":/?=&%+#;,@!$'()*[]~-._")


def has_credentials(source: dict) -> bool:
    header = source.get("header") or {}
    if isinstance(header, str):
        try:
            header = json.loads(header)
        except (ValueError, TypeError):
            return any(word in header.lower() for word in ("cookie", "authorization", "token", "api-key"))
    return isinstance(header, dict) and any(
        any(word in str(key).lower() for word in ("cookie", "authorization", "token", "api-key"))
        for key in header
    )


def read_response(response, sink, deadline):
    if isinstance(response, urllib.error.HTTPError):
        response = response.fp
    # urllib HTTP(S) wraps the socket here; read1 returns after one socket read.
    transport_socket = response.fp.raw._sock
    while not response.isclosed():
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise TimeoutError("response deadline exceeded")
        transport_socket.settimeout(remaining)
        chunk = response.read1(65536)
        if not chunk:
            return True
        accepted = sink(chunk)
        if isinstance(accepted, tuple):
            accepted = accepted[0]
        if not accepted:
            return False
    return True


def successful_probe(live, *, reader_required=False, discovery=False):
    if live.get("status") != "completed" or any(step.get("status") not in {"passed", "success"} for step in live.get("steps", [])):
        return False
    if reader_required:
        document = live.get("sample", {}).get("reader_document", {})
        if document.get("status") != "passed" or document.get("format") != "html" or document.get("reader_entry") is not True:
            return False
    if discovery and (not live.get("detail_intro_bytes") or not live.get("detail_has_cover")
                      or not live.get("cover_image_signature")):
        return False
    return True


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--source-json", required=True, type=Path)
    parser.add_argument("--report", required=True, type=Path)
    parser.add_argument("--online", action="store_true")
    parser.add_argument("--indices", help="one-based comma-separated source indices")
    parser.add_argument("--limit", type=int, default=12)
    parser.add_argument("--keyword")
    parser.add_argument("--discovery", action="store_true", help="probe categories, first category page, details and one cover")
    parser.add_argument("--timeout", type=int, default=10)
    parser.add_argument("--max-requests", type=int, default=8, help="per-source request ceiling for multi-page catalogs, at most 160")
    parser.add_argument("--catalog-pages", type=int, help="limit catalog pages to exercise quick-start parsing")
    parser.add_argument("--catalog-chapters", type=int, help="limit initial chapter metadata, as in first-book startup")
    parser.add_argument("--reader-output-root", type=Path, help="write parsed chapter HTML through the production reader cache")
    parser.add_argument("--require-success", action="store_true", help="exit nonzero unless every selected online flow completes, including requested HTML or discovery metadata")
    args = parser.parse_args()
    if not 1 <= args.timeout <= 20 or not 1 <= args.limit <= 100:
        parser.error("timeout must be 1..20; limit must be 1..100")
    if not 1 <= args.max_requests <= 160 or (args.catalog_pages is not None and not 1 <= args.catalog_pages <= 64):
        parser.error("max-requests must be 1..160; catalog-pages must be 1..64")
    if args.catalog_chapters is not None and not 1 <= args.catalog_chapters <= 10000:
        parser.error("catalog-chapters must be 1..10000")
    if args.require_success and not args.online:
        parser.error("--require-success needs --online")
    if args.discovery and args.reader_output_root:
        parser.error("probe discovery and reader document output in separate runs")
    raw = args.source_json.read_bytes()
    collection = json.loads(raw.decode("utf-8-sig"))
    if not isinstance(collection, list):
        collection = [collection]
    by_url = {source["bookSourceUrl"].strip(): source for source in collection}
    lua = LuaRuntime(unpack_returned_tuples=True, encoding=None)
    module_path = ";".join((ROOT / "legado.koplugin" / suffix).as_posix() for suffix in ("?.lua", "?/init.lua"))
    lua.execute(b"package.path = " + json.dumps(module_path).encode() + b" .. ';' .. package.path")
    lua.globals()[b"probe_zlib_path"] = os.environ.get("LEGADO_ZLIB", "C:/Program Files/Git/mingw64/bin/zlib1.dll" if os.name == "nt" else "libz.so.1").encode()
    lua.execute(b'local ffi=require("ffi"); ffi.loadlib=function(name) assert(name=="z"); return ffi.load(probe_zlib_path) end')
    opener = urllib.request.build_opener(NoRedirect())
    requests = []
    budget_exhausted = False

    def fetch(request, sink):
        nonlocal budget_exhausted
        if len(requests) >= args.max_requests:
            budget_exhausted = True
            return None, b"probe request budget reached"
        url = request[b"url"].decode("utf-8")
        item = {"site": safe_site(url), "method": request[b"method"].decode("ascii")}
        requests.append(item)
        started = time.monotonic()
        deadline = started + float(request[b"timeout"])
        try:
            url = public_url(url)
            headers = {key.decode(): value.decode() for key, value in request[b"headers"].items()}
            headers.setdefault("User-Agent", "Mozilla/5.0 (compatible; KOReader-source-check)")
            headers["Accept-Encoding"] = "identity"
            operation = urllib.request.Request(url, data=request[b"body"], headers=headers, method=item["method"])
            try:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise TimeoutError("request deadline exceeded")
                response = opener.open(operation, timeout=remaining)
            except urllib.error.HTTPError as error:
                response = error
            with response:
                item["http_status"] = response.status
                item["content_type"] = response.headers.get("Content-Type", "")
                item["content_encoding"] = response.headers.get("Content-Encoding", "identity")
                response_headers = {}
                for name in response.headers:
                    values = response.headers.get_all(name)
                    response_headers[name.lower().encode()] = lua.table_from([v.encode() for v in values]) if name.lower() == "set-cookie" else values[-1].encode()
                if not read_response(response, sink, deadline):
                    return None, b"response limit reached"
                return lua.table_from({b"status": response.status, b"headers": lua.table_from(response_headers)}), None
        except Exception as error:
            item["transport_error"] = type(error).__name__
            if isinstance(error, urllib.error.URLError):
                item["transport_cause"] = type(error.reason).__name__
            return None, type(error).__name__.encode()
        finally:
            item["duration_ms"] = round((time.monotonic() - started) * 1000)

    def convert(value, source, target):
        try:
            return value.decode(source.decode()).encode(target.decode())
        except (UnicodeError, LookupError):
            return None, b"character conversion failed"

    reader_root = args.reader_output_root.resolve() if args.reader_output_root else None
    if reader_root:
        reader_root.mkdir(parents=True, exist_ok=True)

    def ensure_reader_paths(source_id, book_id):
        source = source_id.decode("ascii")
        book = book_id.decode("ascii")
        for kind in ("chapters", "html"):
            (reader_root / source / book / kind).mkdir(parents=True, exist_ok=True)

    probe = lua.execute((ROOT / "scripts" / "probe_sources.lua").read_bytes())[b"new"](
        raw, fetch, convert, time.monotonic, args.timeout,
        reader_root.as_posix().encode() if reader_root else None,
        ensure_reader_paths if reader_root else None, args.catalog_pages, args.catalog_chapters)
    inventory = json.loads(probe[b"inventory"]())
    status_counts = Counter(row["status"] for row in inventory)
    requested = {int(item) for item in args.indices.split(",")} if args.indices else None
    report = {"checked_at": datetime.now(timezone.utc).isoformat(),
              "collection_sha256": hashlib.sha256(raw).hexdigest(), "input_entries": len(collection),
              "imported_entries": len(inventory), "static_status_counts": dict(status_counts),
              "transport": "desktop urllib; production Lua RequestEngine normalization/response, parser and Diagnostics",
              "sources": []}
    tested = 0
    sites = set()
    for row in inventory:
        source = by_url[row["url"]]
        site = safe_site(row.pop("url"))
        row["site"] = site
        eligible = row["status"] != "unsupported" and row["enabled"] and not has_credentials(source)
        selected = row["index"] in requested if requested else eligible and site not in sites
        row["live"] = {"status": "not_tested"}
        if args.online and selected and tested < args.limit:
            if has_credentials(source):
                row["live"] = {"status": "skipped_credentials"}
            else:
                tested += 1
                sites.add(site)
                requests.clear()
                budget_exhausted = False
                check_rule = source.get("ruleSearch")
                keyword = args.keyword or (check_rule.get("checkKeyWord") if isinstance(check_rule, dict) else None) or "三国"
                try:
                    row["live"] = json.loads(probe[b"explore"](row["index"]) if args.discovery else probe[b"run"](row["index"], keyword.encode()))
                except Exception as error:
                    row["live"] = {"status": "probe_error", "error_type": type(error).__name__}
                row["live"]["requests"] = list(requests)
                if budget_exhausted:
                    row["live"]["status"] = "probe_limited"
                    row["live"]["probe_limit"] = f"{args.max_requests} HTTP requests per source"
                row["live"]["query"] = keyword
                failure = next((step for step in row["live"].get("steps", []) if step["status"] == "failed"), {})
                print(json.dumps({"index": row["index"], "site": site, "status": row["live"]["status"],
                                  "failed_step": failure.get("name"), "error": failure.get("error", {}).get("code")}), flush=True)
        report["sources"].append(row)
        # Incremental evidence survives an interrupted network batch.
        if args.online and row["live"]["status"] != "not_tested":
            args.report.parent.mkdir(parents=True, exist_ok=True)
            args.report.write_text(json.dumps(report, ensure_ascii=False, indent=2), encoding="utf-8")
    if args.require_success:
        checked = [row["live"] for row in report["sources"] if row["live"]["status"] != "not_tested"]
        passed = sum(successful_probe(live, reader_required=reader_root is not None,
                                      discovery=args.discovery) for live in checked)
        report["validation"] = {"status": "passed" if checked and passed == len(checked) else "failed",
                                "passed": passed, "failed": len(checked) - passed}
    args.report.parent.mkdir(parents=True, exist_ok=True)
    args.report.write_text(json.dumps(report, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps({"input_entries": len(collection), "imported_entries": len(inventory),
                      "static": dict(status_counts), "live_attempted": tested,
                      "validation": report.get("validation")}), flush=True)
    if args.require_success:
        return 0 if report["validation"]["status"] == "passed" else 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/probe_sources.lua
local Json = require("legado.lib.json_codec")
local Importer = require("legado.lib.source_importer")
local Rules = require("legado.lib.rule_engine")
local Safe = require("legado.lib.safe_functions")
local Scanner = require("legado.lib.compatibility_scanner")
local Requests = require("legado.lib.request_engine")
local Service = require("legado.lib.book_service")
local Diagnostics = require("legado.lib.diagnostics")

local Probe = {}

function Probe.new(text, fetch, convert, clock, timeout, reader_root, ensure_reader_paths, catalog_pages, catalog_chapters)
    local sources = {}
    local storage = {
        listSources = function() return sources end,
        replaceSources = function(_, values) sources = values; return true end,
        getSource = function(_, id)
            for _, source in ipairs(sources) do if source.id == id then return source end end
        end,
    }
    local imported = Importer:new({ storage = storage }):importJson(text)
    assert(not imported.error and imported.rejected == 0, "source collection cannot be imported")
    local engine = Requests.new({ scheduler = { scheduleIn = function() end },
        transport = { request = function(_, request, sink) return fetch(request, sink) end },
        charset_converter = function(...) return convert(...) end,
        now = function() return clock() end, logger = {}, settings = { timeout = timeout },
    })
    -- Desktop probe uses the production HTTP policy synchronously; Kindle workers are tested separately.
    local requests = { execute = function(_, request, callback)
        local normalized, err = engine:_normalize(request)
        if not normalized then callback(nil, err)
        else
            for name in pairs(normalized.headers) do
                local lower = name:lower()
                if lower:find("cookie", 1, true) or lower:find("authorization", 1, true)
                    or lower:find("token", 1, true) or lower:find("api-key", 1, true) then
                    normalized.headers[name] = nil
                end
            end
            local outcome = engine:_run_work(normalized, clock() + normalized.timeout)
            callback(outcome.response, outcome.error or (outcome.panic and { code = "PROBE_ERROR" }))
        end
        return { cancel = function() return false end }
    end }
    local rules = Rules.new({ json_decoder = Json, html_parser = require("legado.vendor.htmlparser"),
        safe_functions = Safe.functions, url_resolver = Safe.resolve_url })
    local rule_fields, rule_errors = {}, {}
    local function reset_rule_errors(source)
        rule_fields, rule_errors = {}, {}
        for _, field in ipairs({"ruleExplore", "ruleSearch", "ruleBookInfo", "ruleToc", "ruleContent"}) do
            for key, rule in pairs(type(source[field]) == "table" and source[field] or {}) do
                if type(rule) == "string" and tostring(key):match("^[%w_]+$") then rule_fields[rule] = field .. "." .. key end
            end
        end
    end
    for _, method in ipairs({ "parse", "parseElements" }) do
        local implementation = rules[method]
        rules[method] = function(self, input, rule, ...)
            local result, err = implementation(self, input, rule, ...)
            if err then rule_errors[#rule_errors + 1] = { field = rule_fields[rule] or "unknown", code = err.code,
                limit = type(err.details) == "table" and err.details.limit or nil } end
            return result, err
        end
    end
    local service = Service.new({ storage = storage, rule_engine = rules, request_engine = requests,
        url_template = require("legado.lib.url_template").new({ rule_engine = rules }) })
    local diagnostics = Diagnostics.new({ book_service = service, now = function() return clock() end })
    local reader_cache
    if reader_root then
        reader_cache = require("legado.lib.cache_store").new({ root = reader_root })
    end
    local function check_reader_document(source, book, chapter, body)
        if ensure_reader_paths then ensure_reader_paths(book.source_id, book.id) end
        local stored, store_error = reader_cache:writeBody(book.source_id, book.id, chapter, body)
        if not stored then
            return { status = "failed", code = store_error.code,
                stage = store_error.details and store_error.details.stage or "body_write" }
        end
        local bytes, path
        local session = require("legado.lib.reader_session").new({
            cache = reader_cache,
            storage = { putProgress = function() return true end },
            settings = { get = function() return 0 end },
            ui = { openDocument = function(_, candidate, callbacks)
                path = candidate
                local file = io.open(candidate, "rb")
                if not file then return nil, { code = "STORAGE_ERROR", details = { stage = "reader_open" } } end
                bytes = file:read("*a")
                file:close()
                local document = { getProgressFraction = function() return 0 end }
                callbacks.ready(document)
                return document
            end },
        })
        local opened, open_error = session:open(source, book, { chapter }, 1)
        if not opened then
            return { status = "failed", code = open_error.code,
                stage = open_error.details and open_error.details.stage }
        end
        local valid = type(bytes) == "string" and bytes:sub(1, 15):lower() == "<!doctype html>"
            and bytes:sub(-14):lower() == "</body></html>" and path:sub(-5):lower() == ".html"
        return { status = valid and "passed" or "failed", format = valid and "html" or "invalid",
            bytes = type(bytes) == "string" and #bytes or 0, reader_entry = opened ~= nil }
    end
    local sample
    local get_chapters, get_content = service.getChapters, service.getContent
    function service:getChapters(source, book, callback)
        return get_chapters(self, source, book, function(chapters, err, metadata)
            if chapters and #chapters > 0 then
                local seen, duplicates = {}, 0
                for _, chapter in ipairs(chapters) do
                    if seen[chapter.url] then duplicates = duplicates + 1 end
                    seen[chapter.url] = true
                end
                sample.catalog_duplicate_urls = duplicates
                sample.catalog_complete = metadata and metadata.catalog_complete
                sample.first_chapter_title = chapters[1].title
                sample.last_chapter_title = chapters[#chapters].title
                sample.first_chapter_number = tonumber(chapters[1].title:match("%d+"))
                sample.last_chapter_number = tonumber(chapters[#chapters].title:match("%d+"))
            end
            callback(chapters, err, metadata)
        end, { max_pages=catalog_pages, max_chapters=catalog_chapters })
    end
    function service:getContent(source, book, chapter, callback)
        return get_content(self, source, book, chapter, function(content, err, metadata)
            if content then
                sample.content_bytes = #content.content
                if reader_cache then sample.reader_document = check_reader_document(source, book, chapter, content.content) end
            end
            callback(content, err, metadata)
        end)
    end
    local search_errors
    local search = service.search
    function service:search(keyword, ids, page, callback)
        return search(self, keyword, ids, page, function(result, err, metadata)
            search_errors = result and result.errors
            callback(result, err, metadata)
        end)
    end
    return {
        inventory = function()
            local rows = {}
            for index, source in ipairs(sources) do
                local report = Scanner.scan(source)
                local issues = {}
                for _, issue in ipairs(report.issues) do issues[#issues + 1] = { field = issue.field, code = issue.code } end
                rows[index] = { index = index, name = source.bookSourceName,
                    url = source.bookSourceUrl, status = report.status,
                    capabilities = report.capabilities, issues = issues, enabled = source.enabled }
            end
            return Json.encode(rows)
        end,
        explore = function(index)
            local source = assert(sources[index])
            reset_rule_errors(source)
            local report = { status = "completed", steps = {}, rule_errors = rule_errors }
            local function step(name, err, counts)
                report.steps[#report.steps+1] = { name=name, status=err and "failed" or "passed",
                    error=err and {code=err.code} or nil, field_counts=counts }
                if err then report.status="failed" end
            end
            local categories, err = service:exploreCategories(source)
            step("categories",err,{results=#(categories or {})})
            if err then return Json.encode(report) end
            local selected
            for i, category in ipairs(categories or {}) do
                if category.url and not category.error then selected=i; break end
            end
            if not selected then report.status="no_categories"; return Json.encode(report) end
            report.category=categories[selected].title
            service:explore(source.id,selected,1,function(result, failure)
                local groups=result and result.groups or {}
                local covers,intros=0,0
                for _,group in ipairs(groups) do
                    if group.book.cover_url~="" then covers=covers+1 end
                    if group.book.intro~="" then intros=intros+1 end
                end
                step("category_books",failure,{results=#groups})
                report.card_covers,report.card_intros=covers,intros
                if failure or #groups==0 then if not failure then report.status="empty" end; return end
                service:getBookInfo(source,groups[1].book,function(book, info_error)
                    step("detail",info_error)
                    if not book then return end
                    report.detail_intro_bytes=#book.intro
                    report.detail_has_cover=book.cover_url~=""
                    if book.cover_url=="" then return end
                    requests:execute({url=book.cover_url,source_id=book.source_id,max_bytes=2*1024*1024},function(response, cover_error)
                        step("cover",cover_error)
                        if response then
                            local body=response.body or ""
                            report.cover_bytes=#body
                            report.cover_image_signature=body:sub(1,3)=="\255\216\255" or body:sub(1,8)=="\137PNG\r\n\26\n" or body:sub(1,3)=="GIF" or body:sub(9,12)=="WEBP"
                            if not report.cover_image_signature then report.status="failed" end
                        end
                    end)
                end)
            end)
            return Json.encode(report)
        end,
        run = function(index, keyword)
            local report
            search_errors = nil
            sample = {}
            local source = assert(sources[index])
            reset_rule_errors(source)
            diagnostics:run(source, keyword, function(value) report = value end)
            if report then report.search_errors, report.rule_errors, report.sample = search_errors, rule_errors, sample end
            return Json.encode(assert(report, "synchronous probe did not complete"))
        end,
    }
end

return Probe
